STYLE.md

```diff
@@ -0,0 +1,4068 @@
+
+```go
+type options struct {
+  cache  bool
+  logger *zap.Logger
+}
+
+type Option interface {
+  apply(*options)
+}
+
+type cacheOption bool
+
+func (c cacheOption) apply(opts *options) {
+  opts.cache = bool(c)
+}
+
+func WithCache(c bool) Option {
+  return cacheOption(c)
+}
+
+type loggerOption struct {
+  Log *zap.Logger
+}
+
+func (l loggerOption) apply(opts *options) {
+  opts.logger = l.Log
+}
+
+func WithLogger(log *zap.Logger) Option {
+  return loggerOption{Log: log}
+}
+
+// Open creates a connection.
+func Open(
+  addr string,
+  opts ...Option,
+) (*Connection, error) {
+  options := options{
+    cache:  defaultCache,
+    logger: zap.NewNop(),
+  }
+
+  for _, o := range opts {
+    o.apply(&options)
+  }
+
+  // ...
+}
+```
+
+Note that there's a method of implementing this pattern with closures but we
+believe that the pattern above provides more flexibility for authors and is
+easier to debug and test for users. In particular, it allows options to be
+compared against each other in tests and mocks, versus closures where this is
+impossible. Further, it lets options implement other interfaces, including
+`fmt.Stringer` which allows for user-readable string representations of the
+options.
+
+See also,
+
+- [Self-referential functions and the design of options](https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html)
+- [Functional options for friendly APIs](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)
+
+<!-- TODO: replace this with parameter structs and functional options, when to
+use one vs other -->
+
+## Linting
+
+More importantly than any "blessed" set of linters, lint consistently across a
+codebase.
+
+We recommend using the following linters at a minimum, because we feel that they
+help to catch the most common issues and also establish a high bar for code
+quality without being unnecessarily prescriptive:
+
+- [errcheck](https://github.com/kisielk/errcheck) to ensure that errors are handled
+- [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) to format code and manage imports
+- [golint](https://github.com/golang/lint) to point out common style mistakes
+- [govet](https://pkg.go.dev/cmd/vet) to analyze code for common mistakes
+- [staticcheck](https://staticcheck.dev) to do various static analysis checks
+
+### Lint Runners
+
+We recommend [golangci-lint](https://github.com/golangci/golangci-lint) as the go-to lint runner for Go code, largely due
+to its performance in larger codebases and ability to configure and use many
+canonical linters at once. This repo has an example [.golangci.yml](https://github.com/uber-go/guide/blob/master/.golangci.yml) config file
+with recommended linters and settings.
+
+golangci-lint has [various linters](https://golangci-lint.run/usage/linters/) available for use. The above linters are
+recommended as a base set, and we encourage teams to add any additional linters
+that make sense for their projects.
```

Comment:
We have a `.golangci-lint` configuration that we use for this project. Developers should use the configuration we use when linting their code for a contribution. 🙂

Reply:
Added as a comment :+1: